File: C++/Kadane's Algorithm/buy_and_sell_stock.cpp
#include<bits/stdc++.h>
using namespace std;

int buysell(vector<int>arr, int n){
    vector<int>diff(n, 0);
    for(int i=1; i<n; i++){
        diff[i] = arr[i]-arr[i-1];
    }

    for(int i=1; i<n; i++){
        diff[i] = max(diff[i]+diff[i-1], diff[i]);
    }
    return *max_element(diff.begin(), diff.end());
}

int main(){
    int test, n;
    cin>>test;
    while(test--){
        cin>>n;
        vector<int>arr(n);
        for(int i=0; i<n; i++){
            cin>>arr[i];
        }
        cout<<buysell(arr, n)<<'\n';
    }
}

File: Python/Kadane's Algorithm/buy_and_sell_stock.py
def buysell(n, a):
    if(n==0):
        return 0
    diff = [0]*n
    
    for i in range(1, n):
        diff[i] = a[i]-a[i-1]

    for i in range(1, n):
        diff[i] = max(diff[i]+diff[i-1], diff[i])
    return max(diff)

test = int(input())
for i in range(test):
    n = int(input())
    a = list(map(int, input().split()))
    print(buysell(n, a))
